From b76254e236404a5b47ce78ade8aac24fbb32bfd6 Mon Sep 17 00:00:00 2001
From: Qiu Jianlin <jianlin.qiu@intel.com>
Date: Fri, 22 Apr 2022 13:49:21 +0800
Subject: [PATCH] Patch ffmpeg to address security issues.

This is cherry-picked from below ffmpeg commit:

avcodec/utils: don't return negative values in av_get_audio_frame_duration()
author	James Almer <jamrial@gmail.com>
Wed, 21 Jul 2021 12:02:44 +0800 (01:02 -0300)
committer	James Almer <jamrial@gmail.com>
Sat, 7 Aug 2021 08:22:49 +0800 (21:22 -0300)
In some extrme cases, like with adpcm_ms samples with an extremely high channel
count, get_audio_frame_duration() may return a negative frame duration value.
Don't propagate it, and instead return 0, signaling that a duration could not
be determined.

Fixes ticket #9312

Signed-off-by: James Almer <jamrial@gmail.com>
---
 libavcodec/utils.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/libavcodec/utils.c b/libavcodec/utils.c
index af2a4c2a88..52da84297d 100644
--- a/libavcodec/utils.c
+++ b/libavcodec/utils.c
@@ -1777,20 +1777,22 @@ static int get_audio_frame_duration(enum AVCodecID id, int sr, int ch, int ba,
 
 int av_get_audio_frame_duration(AVCodecContext *avctx, int frame_bytes)
 {
-    return get_audio_frame_duration(avctx->codec_id, avctx->sample_rate,
+    int duration = get_audio_frame_duration(avctx->codec_id, avctx->sample_rate,
                                     avctx->channels, avctx->block_align,
                                     avctx->codec_tag, avctx->bits_per_coded_sample,
                                     avctx->bit_rate, avctx->extradata, avctx->frame_size,
                                     frame_bytes);
+    return FFMAX(0, duration);
 }
 
 int av_get_audio_frame_duration2(AVCodecParameters *par, int frame_bytes)
 {
-    return get_audio_frame_duration(par->codec_id, par->sample_rate,
+    int duration = get_audio_frame_duration(par->codec_id, par->sample_rate,
                                     par->channels, par->block_align,
                                     par->codec_tag, par->bits_per_coded_sample,
                                     par->bit_rate, par->extradata, par->frame_size,
                                     frame_bytes);
+    return FFMAX(0, duration);
 }
 
 #if !HAVE_THREADS
-- 
2.35.1.windows.2

